File: cpp/KalkCpp/Shape/shape.cpp
#include "shape.h"
#include <iostream>
#include <string.h>
#include <typeinfo>



Shape::Shape(std::string n, ColorRGB* c): name(n), color(c)
{

}



Shape::Shape(const Shape& s): name(s.name), color(s.color->clone())
{

}


bool Shape::operator ==(const Shape &s) const
{
    if(this == &s)
        return true;
    if(&s != 0 && typeid(*this) == typeid(s))
    {
        return name == s.name && *color==(*s.color);
    }
    return false;
}



Shape& Shape::operator=(const Shape& s)
{
    if(this != &s){
        name = s.name;
        delete color;
        if(s.color){
            color = s.color->clone();
        }else{
            color = nullptr;
        }
    }
    return *this;
}



ColorRGB* Shape::getColor() const
{
    return color;
}



void Shape::setColor(const ColorRGB& newColor)
{
    delete color;
    color = new ColorRGB(newColor);
}



std::string Shape::getName() const
{
    return name;
}



Shape::~Shape(){
    delete color;
    color=nullptr;
}

File: Java/src/Shape/Shape.java
package Shape;

import Color.ColorRGB;

public abstract class Shape {
    private String name;
    private ColorRGB color;

    public Shape(String n) {
        name = n;
        color = new ColorRGB();
    }

    public Shape(String n, ColorRGB c) {
        this(n);
        color = c;
    }

    public Shape(Shape s) {
        name = s.name;
        color = s.color;
    }

    public abstract int getNumberVertex();

    public ColorRGB getColor() {
        return color;
    }

    public void setColor(ColorRGB c) {
        color = c;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj)
            return true;
        if (obj != null && obj.getClass() == getClass()) {
            return name.equals(((Shape) obj).name) && color.equals(((Shape) obj).color);
        }
        return false;
    }

    public String toString() {
        return "Name: " + name + "\n" + "Color: " + color + "\n";
    }
}
